### PR TITLE
Makefile: add faillint to lint target - Issue 2142

### LINKED ISSUE #2142: Incorporate failinit.
Hi, 

I would propose adding https://github.com/fatih/faillint/blob/master/README.md to our 'make lint' rule to avoid some inconsistencies like some packages we would like to not import:

* errors (we use pkg/errors)
* prometheus/tsdb (it was archived and moved to Prometheus module)
* prometheus/prometheus/pkg/testutils (we have our own)

Help wanted 🤗

### PR DESCRIPTION
**Pull request by kadern0**
Fixes #2142 

* [X] I added CHANGELOG entry for this change. 
(I've included a line at the bottom of the Added section)
* [X] Change is not relevant to the end user.

## Changes

Added failint package with the paths specified within the issue description to 'make lint'

## Verification
Manual execution of the command and output verification.

### CODE REVIEW

**Comment by kadern0** [comment]
This is the output of the command:

```
make lint
>> verifying packages being imported
/home/kaderno/go/src/github.com/improbable-eng/thanos/pkg/block/fetcher.go:9:12: package "errors" shouldn't be imported
/home/kaderno/go/src/github.com/improbable-eng/thanos/pkg/receive/hashring.go:8:2: package "errors" shouldn't be imported
/home/kaderno/go/src/github.com/improbable-eng/thanos/pkg/cacheutil/memcached_client_test.go:8:2: package "errors" shouldn't be imported
/home/kaderno/go/src/github.com/improbable-eng/thanos/pkg/query/api/v1_test.go:22:2: package "errors" shouldn't be imported
/home/kaderno/go/src/github.com/improbable-eng/thanos/pkg/runutil/example_test.go:8:2: package "errors" shouldn't be imported
/home/kaderno/go/src/github.com/improbable-eng/thanos/pkg/store/cache/memcached_test.go:8:2: package "errors" shouldn't be imported
/home/kaderno/go/src/github.com/improbable-eng/thanos/pkg/store/storepb/custom_test.go:7:2: package "errors" shouldn't be imported
```
All these packages import "errors", so they should be updated before merging this PR or builds will fail.

File: Makefile [hunk 1]
```diff
@@ -78,6 +78,8 @@ ALERTMANAGER            ?= $(GOBIN)/alertmanager-$(ALERTMANAGER_VERSION)
 MINIO_SERVER_VERSION    ?= RELEASE.2018-10-06T00-15-16Z
 MINIO_SERVER            ?=$(GOBIN)/minio-$(MINIO_SERVER_VERSION)
 
+IMPORTS_TO_AVOID				?=errors,prometheus/tsdb,prometheus/prometheus/pkg/testutils
```
Comment (bwplotka) [inline Makefile]: I think it would be nice to have full modules path. E.g not everything is hosted in `github.com` domain.

Also: 
```suggestion
MODULES_TO_AVOID				?=errors,prometheus/tsdb,prometheus/prometheus/pkg/testutils
```

File: Makefile [hunk 2]
```diff
@@ -283,14 +285,18 @@ web: web-pre-process $(HUGO)
 	# TODO(bwplotka): Make it --gc
 	@cd $(WEB_DIR) && HUGO_ENV=production $(HUGO) --config hugo.yaml --minify -v -b $(WEBSITE_BASE_URL)
 
+.PHONY: failint
+failint:
```
Comment (bwplotka) [inline Makefile]: Please let's add this to `lint` command. We already have too many commands ;p

File: Makefile [hunk 3]
```diff
@@ -283,14 +285,18 @@ web: web-pre-process $(HUGO)
...
+	@echo ">> verifying packages being imported"
+	@faillint -paths $(IMPORTS_TO_AVOID) ./...
```
Comment (bwplotka) [inline Makefile]: What about vendor? I think we should exclude vendor

File: Makefile (hunk 2 above)
Comment (bwplotka) [reply Makefile]: I I know you added this as lint dep, but I  think it might be more explicit when inlined in `lint` command.

File: Makefile (hunk 3 above)
Comment (bwplotka) [inline Makefile]: Also let's install `failint` as we do for other Go binaries.

File: go.mod [hunk 4]
```diff
@@ -66,7 +66,7 @@ require (
 	github.com/onsi/gomega v1.7.1 // indirect
 	github.com/opentracing/basictracer-go v1.0.0
 	github.com/opentracing/opentracing-go v1.1.0
-	github.com/pkg/errors v0.8.1
+	github.com/pkg/errors v0.9.1
```
Comment (bwplotka) [inline go.mod]: Why we upgraded this module on the way? It looks like irrelevant change.

File: Makefile (hunk 2 above)
Comment (kadern0) [reply Makefile]: It makes moving it inside of the `lint` command although then we would have to wait for the whole linting to complete whenever we want to test (not a huge deal).

File: Makefile (hunk 3 above)
Comment (kadern0) [reply Makefile]: I've followed the "Dependency management" section under the contributing guide. I've realized that if I run `make deps` it removes the `faillint` entries from `go.sum` and `go.mod`

File: go.mod (hunk 4 above)
Comment (kadern0) [reply go.mod]: Don't know when this was changed. I thought it was some automatic dependency resolution.

File: Makefile (hunk 3 above)
Comment (kadern0) [reply Makefile]: If we don't want to run it across all folders, we would have to specify the files and directories manually (or a for loop): pkg, cmd, test, scripts and doc.go. Other option could be, opening a PR to the faillint project with this idea.

What do you think?

Comment (bwplotka) [reply Makefile]: ah.. TBH can we do that? Can we ask question on upstream project how to avoid vendor dir?

Comment (bwplotka) [reply Makefile]: Maybe it already support this?

Comment (kadern0) [reply Makefile]: Current version has no option like that. I'm going to ask in the upstream project to see what they recommend.

Comment (kadern0) [reply Makefile]: According to the project developer, faillint excludes the `vendor` directory automatically.

**Comment by daixiang0** [comment]
Would it be better to change pr title as "Makefile: add faillint to lint target"?

**Comment by kadern0** [comment]
> Would it be better to change pr title as "Makefile: add faillint to lint target"?

Yeah, I like it better. Should I keep the reference to the issue No.?

File: Makefile (hunk 2 above)
Comment (bwplotka) [reply Makefile]: I think that's ok (:

Comment (bwplotka) [reply Makefile]: You can put it in front of it if you think it's fast.

File: Makefile (hunk 3 above)
Comment (bwplotka) [reply Makefile]: Thanks for checking :heart:

Comment (bwplotka) [reply Makefile]: And that's ok - check the binaries like `$(MINIO_SERVER)` and how it install it and version. (: Can we mimic that?

File: go.mod (hunk 4 above)
Comment (bwplotka) [reply go.mod]: Let's remove then (:

File: CHANGELOG.md [hunk 5]
```diff
@@ -678,3 +678,4 @@ Initial version to have a stable reference before [gossip protocol removal](/doc
 - Bucket commands.
 - Downsampling support for UI.
 - Grafana dashboards for Thanos components.
+- Failint for imported packages verification.
```
Comment (bwplotka) [inline CHANGELOG.md]: I think we should not touch 0.1.0 release  ;p

Also this is a tooling, not really relevant for Thanos users, so we don't need changelog entry. (:

File: Makefile (hunk 2 above)
Comment (kadern0) [reply Makefile]: I think I've fixed this, please let me know if you were thinking something different when you said to put it in front.

File: Makefile (hunk 3 above)
Comment (kadern0) [reply Makefile]: Done (but check it, please)!

File: go.mod (hunk 4 above)
Comment (kadern0) [reply go.mod]: Removed

File: CHANGELOG.md (hunk 5 above)
Comment (kadern0) [reply CHANGELOG.md]: I was following the guide :) Removed!

Comment (kadern0) [reply CHANGELOG.md]: About not touching the 0.1.0 version, should I change the PR to a different branch instead of master?

Comment (bwplotka) [reply CHANGELOG.md]: No. Master is fine. Just you put the changelog entry at the bottom, but actually releases goes from newest to oldest.

Anyway, there should be no entry.

**Comment by kadern0** [comment]
One of the tests was failing because of incorrect signing off of commits.  The other test fails because of what I've mentioned in my second comment on this PR. There are 7 packages importing `errors`, thus, the linting stage fails. Now that the "setup" is working, it might be appropriate, fixing these files, what do you think?

Also, yes, I've tested the correct behavior of the code (with different modules and also placing them under a directory called `vendor`) and it seems to be working as expected.

**Comment by kadern0** [comment]
If I update this line:
`MODULES_TO_AVOID        ?=errors,prometheus/tsdb,prometheus/prometheus/pkg/testutils`
to:
`MODULES_TO_AVOID        ?=errors=pkg/errors,prometheus/tsdb,prometheus/prometheus/pkg/testutils`

Then, in the output we get a suggestion of the correct module to import:
`improbable-eng/thanos/pkg/block/fetcher.go:9:12: package "errors" shouldn't be imported, suggested: "pkg/errors"` 

Should we add the recommended modules for each of the ones to avoid?

**Comment by bwplotka** [comment]
That would be amazing! Nice idea @kadern0!

But please let's use the full path for imports if we can?

Looks like we detected errors!
```
 /go/src/github.com/thanos-io/thanos/pkg/block/fetcher.go:9:12: package "errors" shouldn't be imported
/go/src/github.com/thanos-io/thanos/pkg/receive/hashring.go:8:2: package "errors" shouldn't be imported
/go/src/github.com/thanos-io/thanos/pkg/cacheutil/memcached_client_test.go:8:2: package "errors" shouldn't be imported
/go/src/github.com/thanos-io/thanos/pkg/query/api/v1_test.go:22:2: package "errors" shouldn't be imported
/go/src/github.com/thanos-io/thanos/pkg/runutil/example_test.go:8:2: package "errors" shouldn't be imported
/go/src/github.com/thanos-io/thanos/pkg/store/cache/memcached_test.go:8:2: package "errors" shouldn't be imported
/go/src/github.com/thanos-io/thanos/pkg/store/storepb/custom_test.go:7:2: package "errors" shouldn't be imported
```

Do you mind changing those?

Good job, can't wait to merge it! (:

**Comment by kadern0** [comment]
I've changed all the imports but now circleci is failing for some reason I don't understand. Could you please have a look?

File: Makefile [hunk 6]
```diff
@@ -418,6 +424,11 @@ $(ALERTMANAGER):
 $(MINIO_SERVER):
 	$(call fetch_go_bin_version,github.com/minio/minio,$(MINIO_SERVER_VERSION))
 
+$(FAILLINT):
+	@go get github.com/fatih/faillint@$(FAILLINT_VERSION)
+	@mv $(GOBIN)/faillint $(FAILLINT)
+	@go mod tidy
```
Comment (yeya24) [inline Makefile]: Maybe the problem is that the go mod file is changed because of downloading faillint. What about using `fetch_go_bin_version` like we downloading other binaries?

Comment (kadern0) [reply Makefile]: Fixed

File: go.mod (hunk 4 above)
Comment (bwplotka) [reply go.mod]: This is actually needed (:

File: pkg/block/fetcher.go [hunk 7]
```diff
@@ -185,7 +184,7 @@ func (s *MetaFetcher) loadMeta(ctx context.Context, id ulid.ULID) (*metadata.Met
 			return m, nil
 		}
 
-		if !stderrors.Is(err, os.ErrNotExist) {
+		if !errors.Is(err, os.ErrNotExist) {
```
Comment (bwplotka) [inline pkg/block/fetcher.go]: FYI: This actually works thanks to: https://github.com/pkg/errors/commit/6d954f502eb89cd315e4baae5b0e0db516d6f787
